### PR TITLE
Global TokenNetwork (channel manager) contract per token

### PR DESCRIPTION
**Pull request by loredanacirstea**
WIP (discussion points are being added)
Contract restructuring
fixes https://github.com/raiden-network/raiden/issues/1171

! Implementation might change. !
I started integrating most of the changes from the start, to determine what data structures I would need. If necessary, after I have something ok, I can make smaller PRs with the changes.

Open discussion points:
- `settleTimeoutValid`: is 6 blocks enough? uRaiden has a min of 500
- `event TokenNetworkCreated(address token_address, address token_network_address, string token_interface_identifier);` -> e.g. `token_interface_identifier = "ERC20"` -  easier to get all of the erc20-compatible token networks.
- `updateTransfer` callable by anyone - allows `Monitoring Service` to monitor it’s own channels if needed (is this bad?; the participant can use other monitoring services any time)
- `event TransferUpdated(uint channel_identifier, address caller);` - now logs the `msg.sender` (3rd party for ex.) - is there a reason to log the participant on behalf of which the update was made? In the current state, I can only (additionally) log the `partner_address` that provided the signature.
- (note to self) see if a separate `ClosingRequest` structure + mapping makes sense.

Notes on Upgradability & libraries (will come later)
- `TokenNetwork` will contain the core code (library). We will have contracts for each token interface: e.g. `TokenNetworkERC20` 

Notes:
- `ChannelUnlocked` - event argument is `transferred_amount` - the entire amount, not the transfer value.

Settled:
- `closeChannel` - only called by a participant.
- `openChannel`: it is ok to be callable by anyone. Put this info in the specs.
- `deposit` - it won't return anything
- `updateTransfer` & `updateTransferDelegate` with 2 signatures (more details in the future commit)

### CODE REVIEW

File: raiden/smart_contracts/TokenNetwork.sol [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+    }
+
+    modifier timeoutOver(uint channel_identifier) {
+        require(closing_requests[channel_identifier].settle_block_number <= block.number);
+        _;
+    }
+
+    // Note: we use the settleTimeout to check if the channel is open. It must always be > 0.
+    modifier settleTimeoutValid(uint timeout) {
+        require(timeout >= 6 && timeout <= 2700000);
```
Comment (palango) [inline raiden/smart_contracts/TokenNetwork.sol]: Comment where the numbers come from

File: raiden/smart_contracts/TokenNetwork.sol [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+        // Check if the contract is indeed a token contract
+        require(token.totalSupply() > 0);
+    }
+
+    /*
+     *  Public functions
+     */
+
+    /// @notice Opens a new channel between `participant1` and `participant2`.
+    /// Can be called by anyone.
```
Comment (palango) [inline raiden/smart_contracts/TokenNetwork.sol]: The specs don't say if this should be callable by anyone. (however, I don't see issues with that)

File: raiden/smart_contracts/TokenNetwork.sol [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+    function openChannel(
+        address participant1,
+        address participant2,
+        uint settle_timeout)
+        settleTimeoutValid(settle_timeout)
+        public
+        returns (uint)
```
Comment (palango) [inline raiden/smart_contracts/TokenNetwork.sol]: Needs update in the specs

File: raiden/smart_contracts/TokenNetworksRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,34 @@
+pragma solidity ^0.4.17;
+
+import "./Token.sol";
+import "./TokenNetwork.sol";
+import "./Utils.sol";
+
+contract TokenNetworkRegistry is Utils {
```
Comment (palango) [inline raiden/smart_contracts/TokenNetworksRegistry.sol]: Right now we have the following functions for the Registry, do you plan to add them?
- `channelManagerByToken(address token_address)`, would become `getTokenNetworkForToken(...)`
- `channelManagerAddresses()`, would become `getTokenNetworkAddresses()`

File: raiden/smart_contracts/TokenNetwork.sol (hunk 1 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: No, this is a discussion point (added in the description). I just reused what was before.

File: raiden/smart_contracts/TokenNetwork.sol (hunk 2 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Discussed this in the spec doc (comments). But it's good to think it over again. 
I think open + deposit can safely be callable by anyone.

File: raiden/smart_contracts/TokenNetworksRegistry.sol (hunk 4 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetworksRegistry.sol]: In general, the client will get data from events to save gas. At least this was agreed upon during the spec discussions.
- `channelManagerByToken` can now be replaced with `TokenNetworkRegistry.token_to_token_networks(token_address )`
- `channelManagerAddresses` - you get that from the `TokenNetworkCreated` events

Comment (palango) [reply raiden/smart_contracts/TokenNetworksRegistry.sol]: Ah, I missed that `token_to_token_networks` was public. For getting events efficiently, would it make sense to store the contract creation block number somewhere?

Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetworksRegistry.sol]: I don't think this should be done in the SC. You can store/hardcode the block number at which the `TokenNetworkRegistry` was deployed in the Py client and use that for the `TokenNetworkCreated` events -> get the block number for each deployed `TokenNetwork`

File: raiden/smart_contracts/TokenNetwork.sol [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+        uint256 transferred_amount)
+        internal
+    {
+        Channel storage channel = channels[channel_identifier];
+
+        require(channel.participants[participant].initialized);
+        require(nonce > channel.participants[participant].nonce);
+        // Transfers can have 0 value
+        require(transferred_amount >= channel.participants[participant].transferred_amount);
+        require(locksroot != channel.participants[participant].locksroot);
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: This require is incorrect, the locksroot may be '0x0' and it may not change between two balance proofs:

- A balance proof, were there are no pending locks, will have the value "0x0", which is equal the unintialized value.
- A second call to update could be with a direct transfer balance proof, were the transferred_amount increased but the locksroot stayed the same.

File: raiden/smart_contracts/TokenNetwork.sol [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+        require(channels[channel_identifier].participants[partner].initialized);
+
+        Participant storage partner_state = channels[channel_identifier].participants[partner];
+
+        // An empty locksroot means there are no pending locks
+        require(partner_state.locksroot != 0);
+
+        // The lock must not have expired, it does not matter how far in the
+        // future it would have expired
+        require(expiration_block >= block.number);
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: Remove the equal sign.

File: raiden/smart_contracts/TokenNetwork.sol [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+        require(hashlock == keccak256(secret));
+
+        locked_encoded = encodeLock(expiration_block, amount, hashlock);
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: `keccak256(expiration_block, amount, hashlock);` is sufficient here.

File: raiden/smart_contracts/TokenNetwork.sol [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+        computed_locksroot = computeMerkleRoot(locked_encoded, merkle_proof);
+
+        // Note that unlocked locks have to be re-unlocked after a `transferUpdate` with a
+        // locksroot that does not contain this lock.
+        require(partner_state.locksroot == computed_locksroot);
+
+        // A lock can be withdrawn only once per participant
+        key = keccak256(partner_state.nonce, partner_state.locksroot);
+        require(!partner_state.withdrawn_locks[key]);
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: This is incorrect.

The code is allowing only one unlock per merkle root, were we want one unlock per lock in the merkle root.

File: raiden/smart_contracts/TokenNetwork.sol [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+
+        // To account for cases when participant1 does not provide participant2's balance proof
+        // Therefore, participant2's transferred_amount will be lower than in reality
+        participant1_amount = max(participant1_amount, 0);
+
+        // At this point `participant1_amount` is between [0,total_deposit], so this is safe.
+        participant2_amount = total_deposit - participant1_amount;
+
+        // Remove the channel data from storage
+        delete channels[channel_identifier];
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: I'm curious, does this operator works with the `mapping(address => Participant) participants;` in the channel?

File: raiden/smart_contracts/TokenNetwork.sol [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+        // [8:40] amount
+        // [40:72] hashlock
+        assembly {
+            expiration := mload(add(lock, 8))
+            amount := mload(add(lock, 40))
+            hashlock := mload(add(lock, 72))
+        }
+    }
+
+    function computeMerkleRoot(bytes lock, bytes merkle_proof)
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: btw, lock can be renamed to `item` or `entry` and this can be moved to another SC if you want.

File: raiden/smart_contracts/TokenNetwork.sol (hunk 3 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Opened an issue in the spec repo with this.

**Comment by hackaugusto** [comment]
> (note to self) token interface check - either in the registry or in the TokenNetwork contract, no need to do it in both

I suggest in the `TokenNetwork` constructor, because it may be called directly without going through the registry (I'm not saying we will, just that it could, therefor the constructor is a better place for this check)

File: raiden/smart_contracts/TokenNetwork.sol [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+        public
+        returns (bool)
+    {
+        Channel storage channel = channels[channel_identifier];
+
+        // Channel must be open and beneficiary must be one of the participants
+        require(channel.participants[beneficiary].initialized);
+
+        // Channel cannot be closed
+        require(closing_requests[channel_identifier].settle_block_number == 0);
```
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: To be honest, at first I was thinking that it would save gas cost because you only initialize that when you close a channel. Under this assumption we use this in uRaiden.  
But I realized that I have to make some gas tests on `Struct`s. I thought that because they have the types defined, they occupy all storage slots needed from the beginning. I may be wrong and this might happen: if you have a defined variable in the structure but don't initialize it (`closing_address` ..) you don't pay gas.

File: raiden/smart_contracts/TokenNetwork.sol [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+
+        // Sender should have enough balance
+        // TODO: is this necessary? token.transfer should fail anyway if implemented correctly
```
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: yes, agreed. already removed this.

File: raiden/smart_contracts/TokenNetwork.sol [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+
+        return last_channel_index;
+    }
+
+    /// @notice Deposit tokens to an already existent channel.
+    /// Can be called by anyone.
+    function deposit(
+        uint channel_identifier,
+        address beneficiary,
+        uint256 added_amount)
```
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Not sure yet. I'm inclining to leave it like this and just add what you proposed: `openAndDeposit`

File: raiden/smart_contracts/TokenNetwork.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (hackaugusto) [reply raiden/smart_contracts/TokenNetwork.sol]: Yeah, I'm not 100% sure either

File: raiden/smart_contracts/TokenNetwork.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (hackaugusto) [reply raiden/smart_contracts/TokenNetwork.sol]: please consider it, it make idempontent deposits possible.

File: raiden/smart_contracts/TokenNetwork.sol (hunk 9 above)
Comment (LefterisJP) [reply raiden/smart_contracts/TokenNetwork.sol]: I don't see why it wouldn't

File: raiden/smart_contracts/TokenNetworksRegistry.sol (hunk 4 above)
Comment (LefterisJP) [reply raiden/smart_contracts/TokenNetworksRegistry.sol]: WE need to discuss the "getting data from events" idea.

As we can see in practise it does not work for one of the major clients. Parity. Parity can only get events from the block it gets ouf of warp and after. If we say we don't care about that -- then we don't support parity unless it does not use warp.

That's assuming they have not changed this behaviour in latest parity versions.

File: raiden/smart_contracts/TokenNetwork.sol (hunk 9 above)
Comment (hackaugusto) [reply raiden/smart_contracts/TokenNetwork.sol]: how does it knows which keys are being used to actually free the storage?

File: raiden/smart_contracts/TokenNetwork.sol (hunk 7 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: In decodeLock is says the lock has 72 bytes -> it concatenates, not hashes to bytes32 
```
function decodeLock(bytes lock)
        pure
        internal
        returns (uint64 expiration, uint256 amount, bytes32 hashlock)
{
       require(lock.length == 72);
       ....
}
```

File: raiden/smart_contracts/TokenNetworksRegistry.sol (hunk 4 above)
Comment (hackaugusto) [reply raiden/smart_contracts/TokenNetworksRegistry.sol]: I'm am a bit skeptical with this idea TBH

File: raiden/smart_contracts/TokenNetwork.sol (hunk 9 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: It should. I will say for sure after I test it. Until then, this is more of a draft of what should happen.

File: raiden/smart_contracts/TokenNetwork.sol [hunk 14]
```diff
@@ -0,0 +1,568 @@
...
+        uint256 transferred_amount;
+
+        // Value used to order transfers and only accept the latest on calls to
+        // update, this will only be relevant after either #182 or #293 is
+        // implemented.
+        uint64 nonce;
+
+        // This value is set to True after the channel has been opened.
+        // This is an efficient way to mark the channel participants without doing a deposit.
+        // This is uint8 and it gets packed with the nonce.
```
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Yes, that's what I know. I will confirm this after I get tests running and we can see the gas cost. But gas was the reason I chose this implementation. Did not want to store the participant addresses and I needed a way to make sure they are initialized in some way after `openChannel`.

File: raiden/smart_contracts/TokenNetwork.sol (hunk 7 above)
Comment (hackaugusto) [reply raiden/smart_contracts/TokenNetwork.sol]: I don't understand your point.

File: raiden/smart_contracts/TokenNetwork.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Having a hard time atm understanding what you want to say. I view `deposit` as having the `topUp` function. Even if called after the channel has been opened, it does `0 + added_amount`.

File: raiden/smart_contracts/TokenNetwork.sol [hunk 15]
```diff
@@ -0,0 +1,568 @@
...
+        _;
+    }
+
+    modifier stillTimeout(uint channel_identifier) {
+        require(closing_requests[channel_identifier].settle_block_number > block.number);
+        _;
+    }
+
+    modifier timeoutOver(uint channel_identifier) {
+        require(closing_requests[channel_identifier].settle_block_number <= block.number);
```
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Fixed https://github.com/raiden-network/raiden/pull/1286/commits/471d53d5114ef0280ce10ac4b76bc604c19307f2

File: raiden/smart_contracts/TokenNetwork.sol (hunk 5 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: fixed https://github.com/raiden-network/raiden/pull/1286/commits/2bc0095f01937e6b7e854ade02a3ded45a4fba7a

File: raiden/smart_contracts/TokenNetwork.sol (hunk 6 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: fixed https://github.com/raiden-network/raiden/pull/1286/commits/760c004d696f0e6f3a32250ee3139b0975ab599e

File: raiden/smart_contracts/TokenNetwork.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (hackaugusto) [reply raiden/smart_contracts/TokenNetwork.sol]: I assumed topUp was idempontent, renaming is not important **idempotancy is.**

What I'm suggesting is the following:

```
function <name>(uint channel_identifier, address beneficiary, uint156 total_balance) {
        Channel storage channel = channels[channel_identifier];
        Participant storage participant = channel.participants[beneficiary];

        // Channel must be open and beneficiary must be one of the participants
        require(participant.initialized);

        // Channel cannot be closed
        require(closing_requests[channel_identifier].settle_block_number == 0);

        if (participant.balance < total_balance) {
            uint256 added_amount = participant.balance - total_balance;

            // Change the state
            channel.participants[beneficiary].balance += added_amount;

            // Do the transfer
            require(token.transferFrom(msg.sender, address(this), added_amount));

            ChannelNewBalance(channel_identifier, beneficiary, channel.participants[beneficiary].balance);
        }
}
```

File: raiden/smart_contracts/TokenNetwork.sol (hunk 7 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: So, we currently have:
```
(expiration, amount, hashlock) = decodeLock(locked_encoded);
...
h = computeMerkleRoot(locked_encoded, merkle_proof);
```
I now need to create `locked_encoded` from those 3 values. You said a `keccak256` was sufficient there, but a `keccak256` would just be a hash of the 3 values => `bytes32`. And I see that I need to concatenate the values to get 72 bytes

Comment (hackaugusto) [reply raiden/smart_contracts/TokenNetwork.sol]: The merkle tree is composed of the locks hashes:
> I now need to create locked_encoded from those 3 values.

`locked_encoded` is a just the three values concatenated `expiration_block || amount || hashlock`. After it is concatenated together we take the hash of it:

https://github.com/raiden-network/raiden/blob/f32a14713aafea62a6f5b7616c8799ed8f8f6a32/raiden/smart_contracts/NettingChannelLibrary.sol#L439

IOW, the merkle tree is composed of the hash of the encoded locks, and that is equivalente to `keccak256(expiration_block, amount, hashlock);`

> keccak256(...) returns (bytes32): compute the Ethereum-SHA-3 (Keccak-256) hash of the (tightly packed) arguments

https://solidity.readthedocs.io/en/develop/miscellaneous.html#global-variables

Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Argh, did not see the `h = keccak256(lock);` in the `computeMerkleRoot`. Ok, got it.

File: raiden/smart_contracts/TokenNetwork.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Ok, now I know what you mean. I understand the idea, but there are some issues that I see:

1. when "top up" is used, people think about `added_amount` - how much they are going to pay and not about `total_balance` - how much they will have in the end. When you go to top up / recharge your phone card, you don't tell&give the cashier  the `total_balance`, but the `added_amount`.

2. This amount of tokens from the constructor has to also be approved if using an ERC20 token. It's weird to have an approve transaction for `added_amount` and then another channel-related one with `total_balance`.

3. If you use something like ERC223, you have to provide the `added_amount`.

Bottom line, I think it's better to use `added_amount` and make sure the UI is nice and clear for the user and I think it makes more sense when you compare with everyday payments.

[the remark this answers is not in the document]
Comment (hackaugusto) [reply raiden/smart_contracts/TokenNetwork.sol]: 1. Use a different name then.
2. it's doesnt make a different from the constructor perspective, since the initial balance is 0
3. the snippet above is using the `added_amount` for the token call

This is not about UI, it's about a property of the function that is useful. Consider this, you're writing an app that uses raiden's REST api, and the app wants to make a new deposit, but something happens, either the user closes and opens it again, the software crashes, the internet connectiong goes bad, and the REST response is never delivered.

Should the app make a new request? And what will be the excpected total balance if it doesnt, or after the second call?

File: raiden/smart_contracts/TokenNetwork.sol (hunk 8 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: fixed https://github.com/raiden-network/raiden/pull/1286/commits/642fe7432241359ae85408352f18865d4b5f6b6d

File: raiden/smart_contracts/TokenNetwork.sol (hunk 7 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: fixed https://github.com/raiden-network/raiden/pull/1286/commits/7e8cb11449b5b88709c54761f4b274764fe9d8e5

File: raiden/smart_contracts/TokenNetwork.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: So, you want to replace the current `deposit` (added_amount) with a function that receives the `total_balance` as an argument.

(just to have an overview:) A has a balance of 3 tokens and wants to deposit 4 more.

1) ERC20:
- A calls `approve(TokenNetwork, 4)`
- A calls `deposit(channel_id, beneficiary, 7)`

2) ERC223:
- A calls `Token.transfer(TokenNetwork, 4, fallback_data)` -> token calls `TokenNetwork.fallback` -> `TokenNetwork.deposit(channel_id, beneficiary, 4 + beneficiary.balance)`

> the REST response is never delivered. Should the app make a new request?
On the client side you can watch for pending transactions and check A's balance after the transaction in mined. An issue would be network congestion and if you want to resend that transaction with a higher gas cost -> a bit more complicated here to send another transaction with the same nonce as the previous one.

So yes, handling this is easier with a `total_balance` approach because you can just send another separate transaction and not worry about ending up with a higher balance than A wanted.

Does it make sense to have both `topUp(added_amount)` and `setDeposit(total_amount)` in the contract?
Also @LefterisJP , please take a look over this.

[the remark this answers is not in the document]
Comment (LefterisJP) [reply raiden/smart_contracts/TokenNetwork.sol]: So there are 2 questions here.

One is the UI side, which I think we all agree should show the deposit amount you are about to deposit and not the total deposit we will have after deposit call succeeds.

On the API side having an idempotent call is a big plus. So I agree with Augusto that on our API side having a deposit of total amount function call makes sense. So something like `set_total_deposit()`.

> Does it make sense to have both topUp(added_amount) and setDeposit(total_amount)

I would say it does not make sense to have both.

Also the one problem I can see with this idempotency is that you will need to approve the contract to transfer tokens for you. So there you are going to have to use the actual amount of tokens about to top up and not the total.

File: raiden/smart_contracts/SecretRegistry.sol [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+contract SecretRegistry {
+
+    /*
+     *  Data structures
+     */
+
+    string constant public contract_version = "0.3._";
+
+    // Token address => TokenNetwork address
```
Comment (hackaugusto) [inline raiden/smart_contracts/SecretRegistry.sol]: please, fix the comment

File: raiden/smart_contracts/SecretRegistry.sol [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    mapping(bytes32 => uint64) public secret_to_block;
```
Comment (hackaugusto) [inline raiden/smart_contracts/SecretRegistry.sol]: Could you provide a bit of the rationale for the the downcast? e.g.:

> uint64 is sufficient to represents 1494527876342209 years with blocks of 7s

File: raiden/smart_contracts/SecretRegistry.sol [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    function registerSecret(bytes32 secret) public returns (bool) {
+        if (secret_to_block[secret] > 0) {
+            return false;
+        }
+        secret_to_block[secret] = uint64(block.number);
+        SecretRevealed(secret);
+        return true;
+    }
+
+    function getSecretBlockHeight(bytes32 secret) public constant returns (uint64) {
```
Comment (hackaugusto) [inline raiden/smart_contracts/SecretRegistry.sol]: Maybe we should upcast the block number? My idea is to match the `block.number` type.

File: raiden/smart_contracts/TokenNetwork.sol [hunk 19]
```diff
@@ -0,0 +1,697 @@
...
+
+        // Channel cannot be closed
+        require(closing_requests[channel_identifier].settle_block_number == 0);
+
+        require(channel.participants[participant].deposit < total_deposit);
+
+        added_deposit = total_deposit - channel.participants[participant].deposit;
+
+        // Sender should have enough balance
+        require(token.balanceOf(msg.sender) >= added_deposit);
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: I thought we would remove these checks

File: raiden/smart_contracts/TokenNetwork.sol [hunk 20]
```diff
@@ -0,0 +1,697 @@
...
+        require(participant1_state.initialized);
+        require(participant2_state.initialized);
+
+        // Direct token transfers done through the token `transfer` function
+        // cannot be accounted for, these superfluous tokens will be burned,
+        // this is because there is no way to tell which participant (if any)
+        // had ownership over the token.
+        total_deposit = participant1_state.deposit + participant2_state.deposit;
+
+        participant1_amount = (
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: This changed the semantics of the code, the order of payments is not arbitrary, the node that **did not** close the channel must be paid first.

File: raiden/smart_contracts/TokenNetwork.sol [hunk 21]
```diff
@@ -0,0 +1,697 @@
...
+        isClosed(channel_identifier)
+        timeoutOver(channel_identifier)
+        public
+    {
+        uint256 participant1_amount;
+        uint256 participant2_amount;
+        uint256 total_deposit;
+
+        Participant memory participant1_state = channels[channel_identifier].participants[participant1];
+        Participant memory participant2_state = channels[channel_identifier].participants[participant2];
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: I wonder, why are you using `memory` instead of `storage`?

File: raiden/smart_contracts/TokenNetwork.sol [hunk 22]
```diff
@@ -0,0 +1,697 @@
...
+            // 'byte' is not working due to the Solidity parser, so lets
+            // use the second best option, 'and'
+            v := and(mload(add(signature, 65)), 0xff)
+        }
+
+        require(v == 27 || v == 28);
+    }
+
+    // TODO - not used anymore now
+    function decodeLock(bytes lock)
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: This can be removed by now.

File: raiden/smart_contracts/TokenNetwork.sol [hunk 23]
```diff
@@ -0,0 +1,697 @@
...
+    // Channel identifier is a uint256, incremented after each new channel
+    mapping (uint256 => Channel) public channels;
+
+    // Used for determining the next channel identifier
+    // Start from 1 instead of 0, otherwise the first channel will have an additional
+    // 15000 gas cost than the rest
+    uint256 public last_channel_index = 1;
+
+    // The ClosingRequest identfier must be THE SAME as the channel identifier (mapping key) from `channels`
+    mapping (uint256 => ClosingRequest) public closing_requests;
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: What is the verdict? does it save in storage costs to have a separate mapping?

File: raiden/smart_contracts/TokenNetwork.sol [hunk 24]
```diff
@@ -0,0 +1,697 @@
...
+    /// Only a participant may close the channel, providing a balance proof signed by its partner. Callable only once.
+    /// @param channel_identifier The channel identifier - mapping key used for `channels`
+    /// @param nonce Strictly monotonic value used to order transfers.
+    /// @param transferred_amount Total amount of tokens transferred by the channel partner
+    /// to the channel participant who calls the function.
+    /// @param locksroot Root of the partner's merkle tree of all pending lock lockhashes.
+    /// @param additional_hash Computed from the message. Used for message authentication.
+    /// @param signature Partner's signature of the balance proof data.
+    function closeChannel(
+        uint256 channel_identifier,
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: This is missing the token address. It must be added and `require`d, otherwise a balance proof from a different network with the same participants can be used here.

It's also missing the chain_id

File: raiden/smart_contracts/TokenNetwork.sol [hunk 25]
```diff
@@ -0,0 +1,697 @@
...
+    /// Can be called multiple times, only by the non-closing participant.
+    /// @param channel_identifier The channel identifier - mapping key used for `channels`.
+    /// @param nonce Strictly monotonic value used to order transfers.
+    /// @param transferred_amount Total amount of tokens transferred by the channel partner
+    /// to the channel participant who calls the function.
+    /// @param locksroot Root of the partner's merkle tree of all pending lock lockhashes.
+    /// @param additional_hash Computed from the message. Used for message authentication.
+    /// @param closing_signature Signature of the closing participant on the balance proof data.
+    function updateTransfer(
+        uint256 channel_identifier,
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: the token_address is missing here too

File: raiden/smart_contracts/TokenNetwork.sol [hunk 26]
```diff
@@ -0,0 +1,697 @@
...
+    /// @param channel_identifier The channel identifier - mapping key used for `channels`.
+    /// @param nonce Strictly monotonic value used to order transfers.
+    /// @param transferred_amount Total amount of tokens transferred by the channel partner
+    /// to the channel participant who calls the function.
+    /// @param locksroot Root of the partner's merkle tree of all pending lock lockhashes.
+    /// @param additional_hash Computed from the message. Used for message authentication.
+    /// @param closing_signature Closing participant's signature of the balance proof data.
+    /// @param non_closing_signature Non-closing participant signature of the balance proof data.
+    function updateTransferDelegate(
+        uint256 channel_identifier,
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: the token_address is missing here too

File: raiden/smart_contracts/TokenNetwork.sol [hunk 27]
```diff
@@ -0,0 +1,697 @@
...
+    /// @param locked_amount Amount of tokens that the locked transfer values.
+    /// @param hashlock The hash of the secret
+    /// @param merkle_proof The merkle proof needed to compute the merkle root.
+    /// @param secret A value used as a preimage in a HTL Transfer
+    function unlock(
+        uint256 channel_identifier,
+        address partner,
+        uint64 expiration_block,
+        uint256 locked_amount,
+        bytes32 hashlock,
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: the hashlock is not required here, since the secret is provided.

File: raiden/smart_contracts/TokenNetwork.sol [hunk 28]
```diff
@@ -0,0 +1,697 @@
+pragma solidity ^0.4.17;
+
+import "./Token.sol";
+import "./Utils.sol";
+import "./SecretRegistry.sol";
+
+contract TokenNetwork is Utils {
```
Comment (hackaugusto) [inline raiden/smart_contracts/TokenNetwork.sol]: This should probably be renamed to include `ERC20`

File: raiden/smart_contracts/SecretRegistry.sol (hunk 18 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/SecretRegistry.sol]: So you want it `uint256`?

File: raiden/smart_contracts/TokenNetwork.sol (hunk 24 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Yes, the `chain_id` is missing. I will add it.
I used the `TokenNetwork` contract address instead of the token address. This is what we agreed on.

File: raiden/smart_contracts/TokenNetwork.sol (hunk 25 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Same. 
I will add the `chain_id`.
I used the TokenNetwork contract address in `recoverAddressFromSignature`.

File: raiden/smart_contracts/TokenNetwork.sol (hunk 26 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Same. 
I will add the `chain_id`.
I used the TokenNetwork contract address in `recoverAddressFromSignature`.

File: raiden/smart_contracts/SecretRegistry.sol (hunk 18 above)
Comment (hackaugusto) [reply raiden/smart_contracts/SecretRegistry.sol]: depends, does solidity have automatic casting? if no, I think it makes sense.

File: raiden/smart_contracts/TokenNetwork.sol (hunk 25 above)
Comment (hackaugusto) [reply raiden/smart_contracts/TokenNetwork.sol]: you're right, I didn't look inside the `recoverAddressFromSignature` function.

File: raiden/smart_contracts/TokenNetwork.sol (hunk 20 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Why is that? I don't really get it.

File: raiden/smart_contracts/SecretRegistry.sol (hunk 18 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/SecretRegistry.sol]: I will note this and see if there are any storage gas cost optimizations from `uint64`. This is the main reason to downcast, but I am not sure about how mappings are stored. I'll let you know about the result. I've opened a separate issue for this : https://github.com/raiden-network/raiden-contracts/issues/2

File: raiden/smart_contracts/TokenNetwork.sol (hunk 23 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Continuing this here: https://github.com/raiden-network/raiden-contracts/issues/3

File: raiden/smart_contracts/TokenNetwork.sol (hunk 27 above)
Comment (loredanacirstea) [reply raiden/smart_contracts/TokenNetwork.sol]: Yes, totally right. Will change.

**Comment by loredanacirstea** [comment]
Closing this issue.
A new repository for the smart contracts has been created and most of the changes discussed here have already been integrated: https://github.com/raiden-network/raiden-contracts

The rest of the comments have been added as issues in this repository. They will be discussed there from this point on.
